src/common/Smi.Common/Messaging/ControlMessageConsumer.cs

```diff
@@ -59,116 +59,108 @@ public class ControlMessageConsumer : Consumer
+                    return;
+                }
 
-                return;
-            }
+                // Don't pass any unhandled broadcast (to "all") messages down to the hosts
+                if (actor.Equals("all"))
+                    return;
 
-            if (action.StartsWith("ping"))
-            {
-                Logger.Info("Pong!");
-                return;
-            }
+                // Else raise the event if any hosts have specific control needs
+                if (ControlEvent != null)
+                {
+                    Logger.Debug("Control message not handled, raising registered ControlEvent(s)");
+                    ControlEvent(Regex.Replace(action, @"[\d]", ""), body);
 
-            // Don't pass any unhandled broadcast (to "all") messages down to the hosts
-            if (actor.Equals("all"))
-                return;
+                    return;
+                }
 
-            // Else raise the event if any hosts have specific control needs
-            if (ControlEvent != null)
+                // Else we should ignore it?
+                Logger.Warn("Unhandled control message with routing key: " + e.RoutingKey);
+            }
+            catch (Exception exception)
             {
-                Logger.Debug("Control message not handled, raising registered ControlEvent(s)");
-                ControlEvent(Regex.Replace(action, @"[\d]", ""), body);
-
-                return;
+                Fatal("ProcessMessageImpl threw unhandled exception", exception);
             }
-
-            // Else we should ignore it?
-            Logger.Warn("Unhandled control message with routing key: " + e.RoutingKey);
         }
 
-        protected override void ErrorAndNack(IMessageHeader header, BasicDeliverEventArgs deliverEventArgs, string message, Exception exception)
+        /// <summary>
+        /// Ensures the control queue is cleaned up on exit. Should have been deleted already, but this ensures it
+        /// </summary>
+        public override void Shutdown()
         {
-            // Can't Nack the message since it is automatically acknowledged!
-            // This shouldn't really be called for control messages
-            throw new Exception("ErrorAndNack called for control message with routing key: " + deliverEventArgs.RoutingKey);
+            using (IConnection connection = _factory.CreateConnection(_processName + _processId + "-ControlQueueShutdown"))
+            using (IModel model = connection.CreateModel())
+            {
+                Logger.Debug("Deleting control queue: " + ControlConsumerOptions.QueueName);
+                model.QueueDelete(ControlConsumerOptions.QueueName);
+            }
         }
 
+        // NOTE(rkm 2020-05-12) Not used in this implementation
+        protected override void ProcessMessageImpl(IMessageHeader header, IMessage message, ulong tag) => throw new NotImplementedException("ControlMessageConsumer does not implement ProcessMessageImpl");
```

Comment:
Should probably be `NotSupportedException`.  


"Throw a NotSupportedException exception if the implementation of an interface member or an override to an abstract base class method is not possible. For example, the Convert.ToInt32(DateTime) method throws a NotSupportedException exception because no meaningful conversion between a date and time and a 32-bit signed integer exists. The method must be present in this case because the Convert class implements the IConvertible interface."

https://docs.microsoft.com/en-us/dotnet/api/system.notimplementedexception?view=netcore-3.1#notimplementedexception-and-other-exception-types

Reply:
Good point, I'll fix that post-merge. I think we ideally want to sort the inheritance so we don't need to provide an implementation at all though.